### DISCUSSION TITLE
Is Py5 good for display graphics on a headless Pi?

### DISCUSSION THREAD

**Discussion by fallenartist** [py5coding/py5generator#423] [General]
Hello. I just discovered Py5! I'm gathering some hardware for a permanent interactive wall installation that will utilise Raspberry Pi 4. The Pi will be hooked up to a video projector and a speaker. The idea is to display some information and trivia in a loop and interrupt with a sound or a video when a touch button is pressed. Also some stuff streamed from YouTube. Ideally, I'd like for the information to be animated, including typography and graphic elements, so it looks more refined.

I planned to code the graphics with Python and Pygame, among other libraries, but maybe Py5 would be a better solution? What do you think? Many thanks!

**Comment by hx2A** [comment]
Hi, @fallenartist !

First, can you tell me how you found out about py5? I'm curious about what I can do to help grow the community and users.

py5 can work on a headless machine, although my understanding of headless means there is no display of any kind. py5 does require a frame buffer so if there is no physical display, you will need to run a virtual frame buffer instead. Something like:

```bash
/usr/bin/Xvfb :0 -screen 0 1024x768x24 &
export DISPLAY=":0"
```

If the raspberry pi is connected to a video projector, then the projector can serve as the display.

Everything you mentioned sounds achievable with py5. py5 has all of Processing's typography and graphics functionality so you are covered there. For interaction it supports mouse and keyboard events. If your touch button uses some electronics you build and connect to the raspberry pi, then you'd need to write some event code for that, but you'd be doing that anyway with pygame.

I'm not familiar with writing python code to stream movies directly from YouTube but I am sure that is achievable. My only question there is what video resolution you are streaming. 4K might have performance problems but getting that to work well would be a challenge for any library. I would suggest you prototype that part first. If you like I can help you optimize that code for performance to use py5 in the most efficient way possible.

**Comment by fallenartist** [comment]
Many thanks @hx2A !
Well, exactly [this comment](https://github.com/jdf/Processing.py-Bugs/issues/375#issuecomment-1867792986) led me here while researching for possible solutions to use in my case. I've used Processing on Mac and Pi in the past.
By headless Pi I mean Raspberry Pi OS Lite (without GUI). The Pi will be permanently hooked up to a video projector – a HD one, so the 1920x1080 resolution.
The touch event will be done via [this sensor](https://learn.adafruit.com/capacitive-touch-sensors-on-the-raspberry-pi/programming) and GPIO pin.

**Reply by hx2A** [reply]
> Well, exactly [this comment](https://github.com/jdf/Processing.py-Bugs/issues/375#issuecomment-1867792986) led me here while researching for possible solutions to use in my case. I've used Processing on Mac and Pi in the past.

Ah, yes! @villares does an excellent job of planting seeds that lead people to py5.

> By headless Pi I mean Raspberry Pi OS Lite (without GUI)

Oh, I see. It's been a while since I used a Raspberry Pi and forgot they offer a "headless" image to setup a new device.

You'll have to see what is in the headless pi image. I suspect it will be missing the necessary packages for it to drive the display (your projector). Is this correct? Have you gotten the headless image to work with a projector before? If you know how to make a Raspberry Pi OS Lite install work with your projector, you should be good to go.

You'll need to make sure when you install Java, you do not install a headless JVM. The latest version of py5 will give you a clear warning if you install headless Java by mistake.

> a HD one, so the 1920x1080 resolution

Do you want to stream YouTube videos at full screen through py5? I don't remember what a Raspberry Pi is capable of but I would want to test this to make sure you can get a good frame rate.
